components/engine/src/lib.rs

```diff
@@ -211,18 +213,109 @@ enum SeekMode {
     Prefix,
 }
 
+pub struct KeyBuilder {
+    buf: Vec<u8>,
+    start: usize,
+}
+
+impl KeyBuilder {
+    pub fn new(max_size: usize, reserved_prefix_len: usize) -> Self {
+        assert!(reserved_prefix_len < max_size);
+        let mut buf = Vec::with_capacity(max_size);
+        if reserved_prefix_len > 0 {
+            unsafe { buf.set_len(reserved_prefix_len) }
+        }
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn from_vec(vec: Vec<u8>, reserved_prefix_len: usize) -> Self {
+        let buf = if reserved_prefix_len == 0 {
+            vec
+        } else {
+            let mut res = Vec::with_capacity(vec.len() + reserved_prefix_len);
+            unsafe {
+                res.set_len(reserved_prefix_len);
+            }
+            res.extend_from_slice(vec.as_slice());
+            res
+        };
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn from_slice(s: &[u8], reserved_prefix_len: usize) -> Self {
+        let mut buf = Vec::with_capacity(s.len() + reserved_prefix_len);
+        if reserved_prefix_len > 0 {
+            unsafe {
+                buf.set_len(reserved_prefix_len);
+            }
+        }
+        buf.extend_from_slice(s);
+        Self {
+            buf,
+            start: reserved_prefix_len,
+        }
+    }
+
+    pub fn add_key_prefix(&mut self, prefix: &[u8]) {
+        assert!(self.start == prefix.len());
+        unsafe {
+            ptr::copy(prefix.as_ptr(), self.buf.as_mut_ptr(), prefix.len());
+            self.start = 0;
+        }
+    }
+
+    pub fn append(&mut self, content: &[u8]) {
+        assert!(content.len() <= self.buf.capacity() - self.buf.len());
+        self.buf.extend_from_slice(content);
+    }
+
+    pub fn as_ptr(&self) -> *const u8 {
+        unsafe { self.buf.as_ptr().add(self.start) }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        self.buf.len() - self.start
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe { slice::from_raw_parts(self.as_ptr(), self.len()) }
+    }
+
+    pub fn build(self) -> Vec<u8> {
+        if self.start == 0 {
+            self.buf
+        } else {
+            let mut res = Vec::with_capacity(self.len());
```

Comment:
We can do it in place to avoid memory allocation: copy everything after start to position 0 and then return the vector as it is.